tests/30rooms/06membership.pl

```diff
@@ -205,3 +206,157 @@
+
+         make_inviteonly_room($do_request_json)
+         ->on_done( sub {
+            ( $room_id ) = @_;
+         })->then( sub {
+            $do_request_json->(
+               method => "POST",
+               uri    => "/api/v1/rooms/$room_id/invite",
+
+               content => {
+                  id_server => (substr $test_http_server_uri_base, length("https://")),
+                  medium => "email",
+                  address => $invitee_email,
+                  display_name => "Cute things",
+               },
+            ),
+         })->then( sub {
+            $do_request_json->(
+               method => "GET",
+               uri    => "/api/v1/rooms/$room_id/state/m.room.member/$invitee_mxid",
+            )->on_done( sub {
+               my ( $body ) = @_;
+               log_if_fail $body;
+               $body->{membership} eq "invite" or
+                  die "Expected invited user membership to be 'invite'";
+            }),
+         }),
+      );
+   };
+
+sub stub_is_lookup {
+   my ( $email, $mxid, $await_http_request ) = @_;
+   $await_http_request->("/_matrix/identity/api/v1/lookup", sub {
+      my ( $req ) = @_;
+      return unless $req->query_param("medium") eq "email";
+      return unless $req->query_param("address") eq $email;
+      return 1;
+   })->then( sub {
+      my ( $request ) = @_;
+      my $content = ( defined($mxid) ? qq({"medium": "email", "address": "$email", "mxid": "$mxid"}) : qq({}) );
+      $request->respond(make_200($content));
+      Future->done( 1 );
+   })
+};
+
+sub make_200 {
+   my ( $content ) = @_;
+   HTTP::Response->new( 200, "OK", ["Content-Length", length($content)], $content );
+};
+
+sub make_inviteonly_room {
+   my ( $do_request_json ) = @_;
+
+   $do_request_json->(
+      method => "POST",
+      uri    => "/api/v1/createRoom",
+
+      content => {
+         # visibility: "private" actually means join_rule: "invite"
+         # See SPEC-74
+         visibility => "private",
+      },
+   )->then( sub {
+      my ( $body ) = @_;
+      Future->done($body->{room_id});
+   })
+};
+
+test "Can invite unbound 3pid",
+   requires => [qw( do_request_json do_request_json_for more_users test_http_server_uri_base await_http_request )],
+
+   do => sub {
+      my ( $do_request_json, $do_request_json_for, $more_users, $test_http_server_uri_base, $await_http_request ) = @_;
+
+      my $invitee = $more_users->[0];
+      my $invitee_email = "lemurs\@monkeyworld.org";
+      my $room_id;
+
+      # sha256(abc+sha256(123lemurs@monkeyworld.org)
+      # = sha256(abc+377e9ce9132221d02d9c76d0db6fe53f01552c1a7493e5001656882853e60299)
+      # = 16c2f564f9f6ecdc26250d20dfd038198b75da6acef8c6f79b8092f19e8d82fa
+      my $outer_nonce = "abc";
+      my $inner_digest = "377e9ce9132221d02d9c76d0db6fe53f01552c1a7493e5001656882853e60299";
+      my $outer_digest = "16c2f564f9f6ecdc26250d20dfd038198b75da6acef8c6f79b8092f19e8d82fa";
```

Comment:
Rather than hardcode these values you might want to use `Digest::SHA` to compute these. Specifically, `sha256_hex()` is nicely convenient here.